src/object/basic/co_integer48.c

```diff
@@ -0,0 +1,121 @@
+/******************************************************************************
+   Copyright 2020 Embedded Office GmbH & Co. KG
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+******************************************************************************/
+
+/******************************************************************************
+* INCLUDES
+******************************************************************************/
+
+#include "co_core.h"
+
+/******************************************************************************
+* PRIVATE DEFINES
+******************************************************************************/
+
+#define COT_ENTRY_SIZE    (uint32_t)6
+
+/******************************************************************************
+* PRIVATE FUNCTIONS
+******************************************************************************/
+
+static uint32_t COTInt48Size (struct CO_OBJ_T *obj, struct CO_NODE_T *node, uint32_t width);
+static CO_ERR   COTInt48Read (struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size);
+static CO_ERR   COTInt48Write(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size);
+
+/******************************************************************************
+* PUBLIC GLOBALS
+******************************************************************************/
+
+const CO_OBJ_TYPE COTInt48 = { COTInt48Size, 0, COTInt48Read, COTInt48Write, 0 };
+
+/******************************************************************************
+* FUNCTIONS
+******************************************************************************/
+
+static uint32_t COTInt48Size(struct CO_OBJ_T *obj, struct CO_NODE_T *node, uint32_t width)
+{
+    uint32_t result = (uint32_t)0;
+
+    CO_UNUSED(node);
+    CO_UNUSED(width);
+
+    if (CO_IS_DIRECT(obj->Key) != 0) {
+        result = (uint32_t)COT_ENTRY_SIZE;
+    } else {
+        /* check for valid reference */
+        if ((obj->Data) != (CO_DATA)0) {
+            result = (uint32_t)COT_ENTRY_SIZE;
+        }
+    }
+    return (result);
+}
+
+static CO_ERR COTInt48Read(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size)
+{
+    CO_ERR   result = CO_ERR_NONE;
+    uint64_t value;
+
+    CO_UNUSED(node);
+    ASSERT_PTR_ERR(obj, CO_ERR_BAD_ARG);
+    ASSERT_PTR_ERR(buffer, CO_ERR_BAD_ARG);
+
+    if (CO_IS_DIRECT(obj->Key) != 0) {
+        value = (uint64_t)(obj->Data);

```

Comment:
I can't test this, because my biggest microcontroller is a 32bit machine. Therefore directs are limited to 32bit. But it seems to be ok.

Reply:
Oh, are you using a DSP? It's strange you can't use uint64_t. On a ARM or other MCU/MPU gcc or iar manages to combine two 32-bits into uint64_t.
I've tested this on a cortex-M3

Reply:
I mean only the `Direct Storage`, checked with `CO_IS_DIRECT()`. 

You may know: this is a mechanism to save some resources. Normally we use the memory of `obj->data` to store a pointer to the variable. With the so-called `Direct Storage`, we store the value in this pointer memory.

As described in the website, this is possible only for machines which uses pointer width of 64bit. This are (as far as I know) 64bit machines, like Cortex-A54.

Note: On Cortex-M3, you cannot store 64bit values within a pointer. You will loose the upper 32bits.

Reply:
Aah, now I see. It's a intptr_t, so yes, uint32_t on cortex-M3.
I try to think about a workaround for that. It means that I already have some problem by using it.
Need to go deeper then

Reply:
Yes, that's true. But then we need a check of data size before writing/reading to/from using 'Direct Storage'.
Right? Otherwise we risk what you've described above for a data type > 32-bits

Reply:
Ok, I've got this now. So a solution for this could be(under `if(CO_IS_DIRECT(...))`:
```
if (COT_ENTRY_SIZE > sizeof(CO_DATA))
    return -1; // TODO: need to choose correct error here
```
So this way we're safe for all types even on 8/16-bits MCUs. For example AVR8 has sizeof(intptr_t) => 2.

Reply:
Finally I've dealt CO_IS_DIRECT() as error in Integer48. Sorry for the late reply!